x-pack/plugins/stack_connectors/server/connector_types/crowdstrike/crowdstrike.ts

```diff
@@ -0,0 +1,191 @@
+    });
+  }
+
+  public async executeHostActions({ alertIds, ...payload }: CrowdstrikeHostActionsParams) {
+    return this.crowdstrikeApiRequest({
+      url: this.urls.hostAction,
+      method: 'post',
+      params: {
+        action_name: payload.command,
+      },
+      data: {
+        ids: payload.ids,
+        ...(payload.actionParameters
+          ? {
+              action_parameters: Object.entries(payload.actionParameters).map(([name, value]) => ({
+                name,
+                value,
+              })),
+            }
+          : {}),
+      },
+      paramsSerializer,
+      responseSchema: CrowdstrikeHostActionsResponseSchema,
+    });
+  }
+
+  public async getAgentDetails(
+    payload: CrowdstrikeGetAgentsParams
+  ): Promise<CrowdstrikeGetAgentsResponse> {
+    return this.crowdstrikeApiRequest({
+      url: this.urls.agents,
+      method: 'GET',
+      params: {
+        ids: payload.ids,
+      },
+      paramsSerializer,
+      responseSchema: CrowdstrikeGetAgentsResponseSchema,
+    });
+  }
+
+  private async getTokenRequest() {
+    const response = await this.request<CrowdstrikeGetTokenResponse>({
+      url: this.urls.getToken,
+      method: 'post',
+      headers: {
+        accept: 'application/json',
+        'Content-Type': 'application/x-www-form-urlencoded',
+        authorization: 'Basic ' + this.base64encodedData,
+      },
+      responseSchema: CrowdstrikeGetTokenResponseSchema,
+    });
+    const token = response.data?.access_token;
+    if (token) {
+      // Clear any existing timeout
+      clearTimeout(CrowdstrikeConnector.tokenExpiryTimeout);
+
+      // Set a timeout to reset the token after 29 minutes (it expires after 30 minutes)
+      CrowdstrikeConnector.tokenExpiryTimeout = setTimeout(() => {
+        CrowdstrikeConnector.token = null;
+      }, 29 * 60 * 1000);
+    }
+    return token;
+  }
+
+  private async crowdstrikeApiRequest<R extends CrowdstrikeBaseApiResponse>(
+    req: SubActionRequestParams<R>,
+    retried?: boolean
+  ): Promise<R> {
+    try {
+      if (!CrowdstrikeConnector.token) {
+        CrowdstrikeConnector.token = (await this.getTokenRequest()) as string;
+      }
+
+      const response = await this.request<R>({
+        ...req,
+        headers: {
+          ...req.headers,
+          Authorization: `Bearer ${CrowdstrikeConnector.token}`,
+        },
+      });
+
+      return response.data;
+    } catch (error) {
+      if (error.code === 401 && !retried) {
+        CrowdstrikeConnector.token = null;
+        return this.crowdstrikeApiRequest(req, true);
+      }
+
+      throw new Error(error.message);
+    }
+  }
+
+  protected getResponseErrorMessage(
+    error: AxiosError<{ errors: [{ message: string; code: number }] }>
+  ): string {
+    const errorData = error.response?.data?.errors?.[0];
+    if (errorData) {
+      return errorData.message;
+    }
+
+    if (!error.response?.status) {
+      return 'Unknown API Error';
```

Comment:
if there is no status, does that automatically mean there's no additional information? could we stringify the error object maybe to provide additional error context?

Reply:
We should get an error in the check above, but yeah - I think there's not much to add here. I believe we can improve the errors with the next steps though.